File: 2586-longest-square-streak-in-an-array/2586-longest-square-streak-in-an-array.py
from collections import defaultdict
import math

class Solution(object):
    def longestSquareStreak(self, nums):
        n = len(nums)
        ans = 0
        mp = defaultdict(int)
        nums.sort()

        for it in nums:
            root = int(math.sqrt(it))
            if root * root == it:
                mp[it] = mp.get(root, 0) + 1
            else:
                mp[it] = 1
            ans = max(ans, mp[it])
        
        return -1 if ans < 2 else ans
        

File: 2586-longest-square-streak-in-an-array/2586-longest-square-streak-in-an-array.cpp
class Solution {
public:
    int longestSquareStreak(vector<int>& nums) {
        int n = nums.size(), ans = 0;
        if(n == 0) return -1;
        unordered_map<int, int> dp;
        sort(nums.begin(), nums.end());

        for(auto it : nums) {
            int root = sqrt(it);
            if(root * root == it) {
                dp[it] = 1 + dp[root];
            } else {
                dp[it] = 1;
            }
            ans = max(dp[it], ans);
        }

        return ans < 2 ? -1 : ans;
    }
};